### ISSUE TITLE
WithCollectionProperty Icollection

### ISSUE THREAD

**Issue by xavave** [ipvalverde/EPPlus.DataExtractor#21]
I use Entity framework virtual ICollection for proxy creation
how can I use WithCollectionProperty with ICollection please ?

There are constructors for HashSet, List and Collection

If I try to add a new constructor like 

    public ICollectionPropertyConfiguration<TRow> WithCollectionProperty<TCollectionItem>(
            Expression<Func<TRow, ICollection<TCollectionItem>>> propertyCollection,
            string startColumn, string endColumn) where TCollectionItem : class
        {
            var collectionConfiguration = new SimpleCollectionColumnDataExtractor<TRow, ICollection<TCollectionItem>, TCollectionItem>
                (propertyCollection, startColumn, endColumn);

            this.simpleCollectionColumnSetters.Add(collectionConfiguration);

            return this;
        }

but I get an error : ICollection<TCollectionItem> must be a non abstract type with a public parameterless constructor

thank you

**Comment by ipvalverde** [comment]
Hi @xavave, thanks for noticing that one.
Unfortunately, the `SimpleCollectionColumnDataExtractor` creates a new instance of the collection type inside of it. So in this case, it would try to create an insurance of `ICollection<TCollectionItem>`, that is why you see that error.

I'm not on my computer at the moment, I'll try to give a look into it tomorrow on how else this can be done. If you have any suggestions, feel free to share as well.

**Comment by xavave** [comment]
I've added a constructor with ObservableCollection, it should fit my requirements

**Comment by xavave** [comment]
Awesome ! Thank you !
I will take a look at your code to understand how you have made it

**Comment by ipvalverde** [comment]
You're welcome! Just released version 2.1.0 [on Nuget](https://www.nuget.org/packages/EPPlus.DataExtractor/) with this feature

**Comment by xavave** [comment]
Hi Israel,
Can I use an observable Collection with this new implementation ?
I get an error : 
public virtual ObservableCollection<Ingredient> FormulaIngredients { get => formulaIngredients; set => Set(ref formulaIngredients, value); }

error : 
    'Expression of type 'System.Collections.ObjectModel.Collection`1[Formula.DAL.Model.Ingredient]' cannot be used for assignment to type 'System.Collections.ObjectModel.ObservableCollection`1[Formula.DAL.Model.Ingredient]''

and 
.WithCollectionProperty(p => p.FormulaIngredients,
             1, "F", cfg => cfg
              .WithProperty(ing => ing.TempImportId, "RM ID")
              .WithProperty(ing => ing.MainTypeId, "RM type", FnConvertMainTypeName)
              .WithProperty(ing => ing.Weight, "RM poids", fnParseDouble)
             )

**Comment by ipvalverde** [comment]
Could you check if your `formulaIngredients` is null?

If it is null, since ObservableCollection inherits from Collection, the overload used in the `WithCollectionProperty` is the Collection one and then the code tries to create a new instance of Collection<Ingredient> and assign it to your property...

**Comment by ipvalverde** [comment]
Nevermind, I was able to reproduce the issue, it is not related to the property being initialized or not. I should publish a fix by tomorrow.

**Comment by ipvalverde** [comment]
Thanks for noticing that error @xavave. Unfortunately, in order to provide a proper fix, I introduced a breaking change.

**TL;DR**
For your case, of an initialized collection property that is not a `List<T>`, `HashSet<T>` or `Collection<T>`, you can use the new method: `WithInitializedCollectionProperty` instead of `WithCollectionProperty`.

**Full story**
If you are using the method `WithCollectionProperty` for a collection property that is not a `List<T>`, `HashSet<T>` or a `Collection<T>`, an exception will be thrown at runtime. In your case, since your property is of type `ObservableCollection<T>`, the overload that accepts a `Collection<T>` is picked and this generates an error.
What **should** happen in this case, is that the overload with `ICollection<T>`, for initialized collection properties, should be picked instead. To prevent the necessity of specifying generic types when calling the method, now the `WithCollectionProperty` overloads that accept `ICollection<T>` are obsolete, new methods were created specifically for the `ICollection<T>`: `WithInitializedCollectionProperty`

**Comment by xavave** [comment]
thank you, it works with WithInitializedCollectionProperty
